src/backend/metal/src/command.rs

```diff
@@ -2944,100 +2951,65 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
-                                res_offset.vs.samplers += 1;
-                                let out = &mut self.state.resources_vs.samplers[index as usize];
-                                if *out != sampler {
-                                    *out = sampler;
-                                    pre.issue(soft::RenderCommand::BindSampler { stage: pso::Stage::Vertex, index, sampler });
-                                }
+                    let mut stages = [
+                        (pso::Stage::Vertex,   &mut self.state.resources_vs, &info.offsets.vs, &resources.vs),
+                        (pso::Stage::Fragment, &mut self.state.resources_ps, &info.offsets.ps, &resources.ps),
+                    ];
+                    //TODO: remove the checks?
+                    for &mut (stage, ref mut cache, res_offset, pool_offset) in stages.iter_mut() {
+                        for (index, &sampler) in (res_offset.samplers .. )
+                            .zip(&data.samplers[pool_offset.samplers.start as usize .. pool_offset.samplers.end as usize])
+                        {
+                            let out = &mut cache.samplers[index as usize];
+                            if *out != sampler {
+                                *out = sampler;
+                                pre.issue(soft::RenderCommand::BindSampler { stage, index, sampler });
                             }
-                            if layout.stages.contains(pso::ShaderStageFlags::FRAGMENT) {
-                                let index = res_offset.ps.samplers;
-                                res_offset.ps.samplers += 1;
-                                let out = &mut self.state.resources_ps.samplers[index as usize];
-                                if *out != sampler {
-                                    *out = sampler;
-                                    pre.issue(soft::RenderCommand::BindSampler { stage: pso::Stage::Fragment, index, sampler });
-                                }
+                        }
+                        for (index, texture) in (res_offset.textures .. )
+                            .zip(&data.textures[pool_offset.textures.start as usize .. pool_offset.textures.end as usize])
+                        {
+                            let texture = texture.map(|(t, _)| t);
+                            let out = &mut cache.textures[index as usize];
+                            if *out != texture {
+                                *out = texture;
+                                pre.issue(soft::RenderCommand::BindTexture { stage, index, texture });
                             }
                         }
+                        for (index, &buffer) in (res_offset.buffers .. )
+                            .zip(&data.buffers[pool_offset.buffers.start as usize .. pool_offset.buffers.end as usize])
+                        {
+                            cache.buffers[index as usize] = buffer;
```

Comment:
Is there any risk of `cache.buffers` becoming stale when `pre.is_void()` occurs (i.e. we don't submit `BindBuffer`)? What's the reason only `BindBuffer` is optional in that case?

Reply:
So wouldn't `if !pre.is_void() { pre.issue(soft::RenderCommand::BindSampler { stage, index, sampler }); }` be valid, for example?

Reply:
all of the `is_void` checks we have are solely for optimization, they aren't functional. It's just to spare some work done for producing the commands that are passed to `pre.issue`